src/Chainweb/Pact/PactService.hs

```diff
@@ -418,246 +410,177 @@ serviceRequests memPoolAccess reqQ = do
-      psEnv <- ask
-      logger <- view psLogger
-      let validate bhi _bha txs = do
-
-            let parentTime = ParentCreationTime $ _blockCreationTime $ _parentHeader parent
-            results <- do
-                let v = _chainwebVersion psEnv
-                    cid = _chainId psEnv
-                validateChainwebTxs logger v cid cp parentTime bhi txs return
-
-            V.forM results $ \case
-                Right _ -> return True
-                Left _e -> return False
-
-      liftIO $!
-        mpaGetBlock mpAccess bfState validate (pHeight + 1) pHash (_parentHeader parent)
-
-    doNewBlock pdbenv = do
-        logInfo $ "(parent height = " <> sshow pHeight <> ")"
-                <> " (parent hash = " <> sshow pHash <> ")"
-
-        blockGasLimit <- view psBlockGasLimit
-        let initState = BlockFill blockGasLimit mempty 0
-
-        let
-            txTimeHeadroomFactor :: Double
-            txTimeHeadroomFactor = 5
-            -- 2.5 microseconds per unit gas
-            txTimeLimit :: Micros
-            txTimeLimit = round $ (2.5 * txTimeHeadroomFactor) * fromIntegral blockGasLimit
-
-        -- Heuristic: limit fetches to count of 1000-gas txs in block.
-        let fetchLimit = fromIntegral $ blockGasLimit `div` 1000
-
-        newTrans <- getBlockTxs initState
-
-        -- NEW BLOCK COINBASE: Reject bad coinbase, always use precompilation
-        Transactions pairs cb <- execTransactions False miner newTrans
-          (EnforceCoinbaseFailure True)
-          (CoinbaseUsePrecompiled True)
-          pdbenv
-          Nothing
-          (Just txTimeLimit) `catch` handleTimeout
-
-        successes <- liftIO $ Vec.new @_ @_ @(ChainwebTransaction, P.CommandResult [P.TxLogJson])
-        failures <- liftIO $ Vec.new @_ @_ @GasPurchaseFailure
-        BlockFill _ requestKeys _ <- refill fetchLimit txTimeLimit pdbenv successes failures =<<
-          foldM (splitResults successes failures) (incCount initState) pairs
-
-        logInfo $ "(request keys = " <> sshow requestKeys <> ")"
-
-        liftIO $ do
-          txHashes <- Vec.toLiftedVectorWith (\_ failure -> pure (gasPurchaseFailureHash failure)) failures
-          mpaBadlistTx mpAccess txHashes
-
-        !pwo <- liftIO $ do
-          txs <- Vec.toLiftedVector successes
-          pure (toPayloadWithOutputs miner (Transactions txs cb))
-        return $! Discard pwo
-
-    refill :: Word64 -> Micros -> PactDbEnv' logger -> GrowableVec (ChainwebTransaction, P.CommandResult [P.TxLogJson]) -> GrowableVec GasPurchaseFailure -> BlockFill -> PactServiceM logger tbl BlockFill
-    refill fetchLimit txTimeLimit pdbenv successes failures = go
+    -> PactServiceM logger tbl (T2 ParentHeader PayloadWithOutputs)
+execNewBlock mpAccess miner = do
+    latestHeader <- findLatestValidBlockHeader
+    T2 latestHeader <$> execNewBlock' latestHeader
+    where
+
+    -- | Note: The BlockHeader param here is the PARENT HEADER of the new
+    -- block-to-be
+    --
```

Comment:
this is parent because `findLatestValidBlockHeader` will retrieve the latest header from the blockheader db which is by definition the parent block of the current header being built.